src/data-validation.jl

```
@@ -0,0 +1,80 @@
+export DataValidationException
+
+"""
+    DataValidationException
+
+Exception related to data validation of the Tulipa Energy Model input data.
+"""
+mutable struct DataValidationException <: Exception
+    error_messages::Vector{String}
+end
+
+function Base.showerror(io::IO, ex::DataValidationException)
+    println(io, "DataValidationException: The following issues were found in the data:")
+    for error_message in ex.error_messages
+        println(io, "- " * error_message)
+    end
+end
+
+"""
+    validate_data!(connection)
+
+Raises an error if the data is not valid.
+"""
+function validate_data!(connection)
+    error_messages = String[]
+    @timeit to "no duplicate rows" append!(error_messages, _validate_no_duplicate_rows!(connection))
+
+    if length(error_messages) > 0
+        throw(DataValidationException(error_messages))
+    end
+
+    return
+end
```

So, for my understanding, in this function, we can start adding more validation functions, correct?
And second, any validation function must/(should?) return a text (i.e., String) type with a meaningful message to the user, correct?

---

1. Yes
2. In this implementation, yes. There are other ways to handle this. For instance, we can fail faster by making every function throw an exception directly, but then if you have multiple issues, you would need to run it multiple times. So I prefer aggregating the error messages. Also, there might be some tweaks, but let's have some validation functions first then think about them.